### PR TITLE
8256730: Code that uses Object.checkIndex() range checks doesn't optimize well

### PR DESCRIPTION
**Pull request by rwestrel**
This was reported by Paul with the vector API. There are 2 issues:

- CastII nodes (added by Objects.checkIndex()) gets in the way of the
  pattern matching performed by range check elimination

- By transforming (CastII (AddI x y)) into (AddI (CastII x) (CastII y))
  some CastII can be eliminated which improves address computation code.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8256730](https://bugs.openjdk.java.net/browse/JDK-8256730): Code that uses Object.checkIndex() range checks doesn't optimize well

### Reviewers
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to faf980ab7d0cba825b74dc9d0555e44f53f51778
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1342/head:pull/1342`
`$ git checkout pull/1342`

### CODE REVIEW

**Comment by neliasso** [comment]
> By transforming (CastII (AddI x y)) into (AddI (CastII x) (CastII y))
some CastII can be eliminated which improves address computation code.

Wasn't there a recent bugfix that did the opposite? I think it was in a different context but I will try to find out.

**Comment by neliasso** [comment]
>> By transforming (CastII (AddI x y)) into (AddI (CastII x) (CastII y))
some CastII can be eliminated which improves address computation code.

> Wasn't there a recent bugfix that did the opposite? I think it was in a different context but I will try to find out.

False alarm. It was a completely different problem.

File: src/hotspot/share/opto/castnode.cpp [hunk 1]
```diff
@@ -246,6 +246,38 @@ Node *CastIINode::Ideal(PhaseGVN *phase, bool can_reshape) {
...
+      return NULL;
+    }
+    int op = z->Opcode();
+    Node* x = z->in(1);
+    Node* y = z->in(2);
+
+    Node* cx = new CastIINode(x, rx->is_int(),
+                              _carry_dependency, _range_check_dependency);
+    cx->set_req(0, in(0));
+    cx = phase->transform(cx);
```
Comment (robcasloz) [inline src/hotspot/share/opto/castnode.cpp:269]: Hi Roland, have you tested that these calls to `transform()` do not create an explosion of recursive `Ideal()` calls in the face of long AddI chains (see https://github.com/openjdk/jdk/pull/727).

Comment (rwestrel) [reply src/hotspot/share/opto/castnode.cpp:269]: I wondered about that but had no issue during testing. Let me see if I can change your test case to have recursive Ideal() calls.

**Comment by PaulSandoz** [comment]
Code generation is good. If also improves other cases, for example https://github.com/richardstartin/vectorbenchmarks/blob/master/src/main/java/com/openkappa/panama/vectorbenchmarks/BitmapLogicals.java#L69

File: src/hotspot/share/opto/castnode.cpp (hunk 1 above)
Comment (rwestrel) [reply src/hotspot/share/opto/castnode.cpp]: I managed to reproduce a similar issue by tweaking your test cases. I just pushed the new test cases + a change that mirrors your fix.

Comment (robcasloz) [reply src/hotspot/share/opto/castnode.cpp]: Great, thanks!

**Comment by rwestrel** [comment]
Anyone for a review of this one?

File: src/hotspot/share/opto/convertnode.cpp [hunk 2]
```diff
@@ -30,6 +30,7 @@
 #include "opto/phaseX.hpp"
 #include "opto/subnode.hpp"
 #include "runtime/sharedRuntime.hpp"
+#include "compile.hpp"
```
Comment (iwanowww) [inline src/hotspot/share/opto/convertnode.cpp:33]: Redundant or `opto/compile.hpp`?

**Review by iwanowww — APPROVED** [review]
Looks good.

File: src/hotspot/share/opto/convertnode.cpp (hunk 2 above)
Comment (rwestrel) [reply src/hotspot/share/opto/convertnode.cpp]: Right. That one doesn't seem required.

**Comment by PaulSandoz** [comment]
@rwestrel this has now been approved, did you forget about it?

**Comment by rwestrel** [comment]
> @rwestrel this has now been approved, did you forget about it?

I didn't but the rule for non trivial hotspot/compiler changes is 2 reviews unless that has changed.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me as well.

**Comment by rwestrel** [comment]
@iwanowww @TobiHartmann thanks for the reviews

**Comment by PaulSandoz** [comment]
> > @rwestrel this has now been approved, did you forget about it?
> 
> I didn't but the rule for non trivial hotspot/compiler changes is 2 reviews unless that has changed.

Ah, my apologies, i was mislead by "This change now passes all automated pre-integration checks."

**Pull request closed without merging**